### PR TITLE
Interp edits

### PR DESCRIPTION
**Pull request by gabi-250**
It can now interpret very simple programs.

### CODE REVIEW

File: interpreter.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+jit_driver = jit.JitDriver (
+    greens = [],
+    reds = [])
+
+class Interpreter(object):
+
+    def __init__(self, global_state = None):
+        self.global_state = global_state
+
+    # only works for integers
```
Comment (vext01) [inline interpreter.py:41]: This part is OK for now, but my gut feeling says you will need to re-engineer this part later.

Comment (vext01) [inline interpreter.py:41]: You may need a _get_int_arg and a get_str_arg method for example...

**Comment by gabi-250** [comment]
Many other things will need to be re-engineered

File: interpreter.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            x, y = self._get_args(args, state)
+            return x + y
+        elif opcode == LLVMMul:
+            x, y = self._get_args(args, state)
+            return x * y
+        elif opcode == LLVMCall:
+            string_format_ref = LLVMGetOperand(args[0], 0)
+            string_format = None
+            # need to access @.str value
+            if state.has_key(rffi.cast(rffi.INT, string_format_ref)):
```
Comment (vext01) [inline interpreter.py]: Does it make sense to abstract variable lookups into a `lookup_var` method (perhaps on your state object), which checks first locals, then globals? Then you wouldn't need to duplicate the logic each time (e.g. also duplicated in `_get_args`).

Comment (gabi-250) [inline interpreter.py]: Yes, I've thought about doing something similar. I'll make that change now

File: interpreter.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                printf_args = []
+                for arg in args[1:-1]:
+                    if state.has_key(rffi.cast(rffi.INT, arg)):
+                        printf_args.append(state.get_variable(rffi.cast(rffi.INT, arg)))
+                    elif self.global_state.has_key(rffi.cast(rffi.INT, arg)):
+                        printf_args.append(self.global_state.get_variable(rffi.cast(rffi.INT, arg)))
+                    elif LLVMIsConstant(arg):
+                        printf_args.append(LLVMConstIntGetSExtValue(arg))
+                    else:
+                        # raise custom exception
```
Comment (vext01) [inline interpreter.py]: This part is not implemented right? If so, raise `NotImplementedError` after printing a message so as to stop the interpreter?

File: interpreter.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+            string_format = None
+            # need to access @.str value
+            if state.has_key(rffi.cast(rffi.INT, string_format_ref)):
+                string_format = state.get_variable(rffi.cast(rffi.INT, string_format_ref))
+            elif self.global_state.has_key(rffi.cast(rffi.INT, string_format_ref)):
+                string_format = self.global_state.get_variable(rffi.cast(rffi.INT, string_format_ref))
+            else:
+                print "error"
+            fn_name = rffi.charp2str(LLVMGetValueName(args[-1]))
+            if fn_name == "printf":
```
Comment (vext01) [inline interpreter.py:67]: I guess at some point you will need a method lookup table or something similar?

File: interpreter.py (hunk 3 above)
Comment (gabi-250) [inline interpreter.py]: Will do

File: interpreter.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                    elif LLVMIsConstant(arg):
+                        printf_args.append(LLVMConstIntGetSExtValue(arg))
+                    else:
+                        # raise custom exception
+                        print "error"
+                print string_format % tuple(printf_args)
+            else:
+                raise NotImplementedError(fn_name)
+        elif opcode == LLVMAlloca:
+            pass
```
Comment (vext01) [inline interpreter.py]: If these are not implemented, print a message and raise `NotImplementedError`? Perhaps make a function to do that, e.g. `def exit_not_implemented(self, message): ...`

File: interpreter.py (hunk 4 above)
Comment (gabi-250) [inline interpreter.py:67]: Yep, but I'm not sure how to implement it yet. If you want I can try to do it now.. or should I just leave it for later?

File: interpreter.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        main_fun = LLVMGetNamedFunction(module, "main")
+
+    with lltype.scoped_alloc(rffi.VOIDPP.TO, 1) as basic_blocks_main_ptr:
+        LLVMGetBasicBlocks(main_fun, basic_blocks_main_ptr)
+        basic_blocks_main = basic_blocks_main_ptr[0]
+
+    global_state = State()
+    global_var = LLVMGetFirstGlobal(module)
+
+    # assuming all global_vars are strings
```
Comment (vext01) [inline interpreter.py]: Is this a safe assumption? Should be have an assertion or exit with a message if we find a non-string?

File: llvm_wrapper.py [hunk 7]
```diff
@@ -5,45 +5,119 @@
 from rpython.translator.tool.cbuild import ExternalCompilationInfo
 
 class CConfig(object):
+
```
Comment (vext01) [inline llvm_wrapper.py]: Blank line

File: interpreter.py (hunk 6 above)
Comment (gabi-250) [inline interpreter.py]: It's not, I'll add a check.

File: state.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    def __init__(self):
+        self.vars = []
+        self.var_offsets = {}
+
+    def get_variable(self, name):
+        ''' Returns the value of the specified local variable. '''
+        off = self.get_variable_off(name)
+        return self.vars[off]
+
+    #@jit.elidable_promote()
```
Comment (vext01) [inline state.py]: commented code.

File: state.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    def get_variable_off(self, name):
+        ''' Returns the offset of the specified local variable. '''
```
Comment (vext01) [inline state.py:16]: Nitpicking, but people usually put a blank line after a docstring.

File: interpreter.py (hunk 4 above)
Comment (vext01) [inline interpreter.py:67]: This can come later. Just seeing if you have thought about it.

**Comment by vext01** [comment]
A few comments there. Overall this is in great shape. Good job.

Fix the comments and we can think about squashing.

**Comment by vext01** [comment]
Does it translate btw?

**Comment by gabi-250** [comment]
I'll make those changes. I haven't tried running it translated in a while. I'll try now.

**Comment by gabi-250** [comment]
Ah, it doesn't... I'm probably mixing different types in a list or something. Should I get it to translate first and then fix the other bugs?

**Comment by vext01** [comment]
Up to you.

**Comment by gabi-250** [comment]
I'll leave it for later, it is a bit complicated because I have to rethink my variable dictionary: it seems that I'm mapping variables both to strings and integers...

**Comment by ltratt** [comment]
Although we can let it go this time, if need be, in general, we should only be committing stuff that definitely translates. The reason is partly pragmatics: the longer things go without being translated, the harder it is to eventually get them to translate.

**Comment by gabi-250** [comment]
Point taken

**Comment by vext01** [comment]
Ah, your symbol table problem is interesting.

You have a couple of options:
- Store raw LLVM pointer instead of converting them to RPython strings and ints.
- Make some kind of wrapper class, which wraps LLVM pointers and has sub-types for ints and strings.

Hrm.

**Comment by gabi-250** [comment]
I'm not sure about the first one. I don't think it would work, given the fact that I also store values which are the result of various operations, such as 'add'.

**Comment by vext01** [comment]
Anyway, we won't let this block this PR, but after this is merged, have a think about that.

**Comment by vext01** [comment]
Looks good. Would you like to squash?

**Comment by gabi-250** [comment]
Sure.

**Comment by vext01** [comment]
This was merged separately in another PR due to git issues. Closing.

**Pull request closed without merging**